### ISSUE TITLE
Live stream very garbled, slow stream

### ISSUE THREAD

**Issue by BigDogJonx** [KhaosT/homebridge-camera-ffmpeg#90]
Here is an example of the issue Im having.  I have this RTSP coming from my QSee NVR.  Its only pulling the lower quality stream, not the higher quality 4MP stream.  I have a Raspberry Pi 3b and using this plugin (as I seem to get the best results of others).  I have ffmpeg compiled using the Raspberry Pi OMX version of it based on this link here (https://www.npmjs.com/package/homebridge-camera-ffmpeg-omx), just step1 there.  My CPU on the RPi definitely isn't pegged or even going higher than 25% with one stream.  If I wait about 10-15secs, the image finally gets clear, but the stream isn't what I call live, its showing me basically one frame maybe every 5-10secs.  

Ive tried countless number of different ffmpeg builds, and the one mentioned above only gives me the best results (although the stream still sucks basically).  Using libav as part of RPi doesn't even work, There was mention of basically installing libav-tools using apt-get and then just creating an alias for ffmpeg, but that doesn't work with these plugins I find, it just crashes homebridge the moment a stream is pulled up on homekit.

Ive also tried this on a much more powerful computer (installed turnkey Linux) and same results.  Nothing really improved, so im confident it isn't the the RPi we have.  

Here is an example of the config.json snippit of the cameras:

`			"platform": "Camera-ffmpeg",

			"cameras": [{
					"name": "FrontPorch",
					"videoConfig": {
						"source": "-re -i rtsp://192.168.90.30:554/cam/realmonitor?channel=1&subtype=01&authbasic=masldmlkamsdlmsalml=",
						"stillImageSource": "-i http://10.0.19.10/Streaming/channels/101/picture",
						"maxStreams": 2,
						"maxWidth": 704,
						"maxHeight": 480,
						"maxFPS": 20,
						"vcodec": "h264_omx"
					}
				},
				{
					"name": "GarageDriveway",
					"videoConfig": {
						"source": "-re -i rtsp://192.168.90.30:554/cam/realmonitor?channel=2&subtype=01&authbasic=masldmlkamsdlmsalml=",
						"stillImageSource": "-i http://10.0.19.10/Streaming/channels/101/picture",
						"maxStreams": 2,
						"maxWidth": 704,
						"maxHeight": 480,
						"maxFPS": 20,
						"vcodec": "h264_omx"
					}
				},
				{
					"name": "Driveway",
					"videoConfig": {
						"source": "-re -i rtsp://192.168.90.30:554/cam/realmonitor?channel=3&subtype=01&authbasic=masldmlkamsdlmsalml=",
						"stillImageSource": "-i http://10.0.19.10/Streaming/channels/101/picture",
						"maxStreams": 2,
						"maxWidth": 704,
						"maxHeight": 480,
						"maxFPS": 20,
						"vcodec": "h264_omx"
					}
				},
				{
					"name": "Garage",
					"videoConfig": {
						"source": "-re -i rtsp://192.168.90.30:554/cam/realmonitor?channel=4&subtype=01&authbasic=masldmlkamsdlmsalml=",
						"stillImageSource": "-i http://10.0.19.10/Streaming/channels/101/picture",
						"maxStreams": 2,
						"maxWidth": 704,
						"maxHeight": 480,
						"maxFPS": 20,
						"vcodec": "h264_omx"
					}
				},`

![img_8141](https://cloud.githubusercontent.com/assets/29067981/26594184/af84b8ec-4534-11e7-9192-40914b7308fd.PNG)

**Comment by BigDogJonx** [comment]
Just wanted to also add that during a live stream, here is what I see on homebridge, its definitely using the -vcodec h264_omx as you can see. 

`-re -i rtsp://192.168.90.30:554/cam/realmonitor?channel=9&subtype=01&authbasic=lkaslkdmlsamdlkma= -threads 0 -vcodec h264_omx -an -pix_fmt yuv420p -r 20 -f rawvideo -tune zerolatency -vf scale=640:360 -b:v 132k -bufsize 132k -payload_type 99 -ssrc 1 -f rtp -srtp_out_suite AES_CM_128_HMAC_SHA1_80 -srtp_out_params XxZxsUc3OzxcQNn9CObpxcFA8RqElnAakTbSLXl/xbUe4Vb srtp://192.168.90.235:63942?rtcpport=63942&localrtcpport=63942&pkt_size=1378
`

**Comment by jsmeyers** [comment]
I have tried decreasing the "pkt_size" to 378 in FFMPEG.js in line 251 with positive results for streaming from my RPi3. Streams start up faster, decrease choppy video. I have been meaning to ask the exact reasoning behind the use of a packet size of 1378 packet except to prevent a packet being generated that would be too large for a VPN or for higher quality cameras. 

Here is my command:

-re -timelimit 45 -i rtsp://10.19.27.47/video.mp4 -threads 2 -vcodec h264_omx -an -pix_fmt yuv420p -r 15 -f rawvideo -tune zerolatency -vf scale=640:360 -b:v 132k -bufsize 132k -payload_type 99 -ssrc 1 -f rtp -srtp_out_suite AES_CM_128_HMAC_SHA1_80 -srtp_out_params Kh3PEDJSkXJAxAsqxWqnL0YiOM+hSSmMPUzLX2mP srtp://10.19.27.30:63305?rtcpport=63305&localrtcpport=63305&pkt_size=378

**Comment by BigDogJonx** [comment]
I wasn't able to get much difference, the garbled video look still occurs but disappears quicker now (like 5-10secs).  What I noticed on a thread on ffmpeg forums was the number for the pkt size should be multiples of 188, they recommend 1316 for the size.  Here is the thread and snippit of the info.

> To make ffmpeg stream workable with commonly used decoders, output needs to be multiple of 188 bytes. MTU is 1500 is standart. So, 7x118 = 1316 bytes. If a part of ts package added to left space on MTU, decoders loose sync or you face with pixels/drop frames. Thats why best practice is to set payload with 1316 bytes with padding null. 

https://trac.ffmpeg.org/ticket/2748

So I set it to 1316, very slightly bit better, but still not what Id call a stream, it shows images maybe every 2-5secs, the RPi3 is only at maybe 8-10% CPU.  I also changed all FPS on the config.json to 30fps, saw in another thread here on this plugin that they recommend it to be 30 otherwise apple homekit apparently has issues. Even if the camera doesn't support it, was said to set it to 30, so I did that as well.

Dixit

**Comment by theo-69** [comment]
@BigDogJonx @jsmeyers  Having the same problem, did you find any solution?

**Comment by BigDogJonx** [comment]
Ive really gotten no improvement, setting is still at 1316 and CPU still aint anywhere near pegging.  30fps was the other thing I put in Config.json but again its still no where near realtime or even showing 5-10fps, its barely showing 1-2fps.

Dixit

**Comment by a900927** [comment]
fps=0.01
anybody help

**Comment by fulvio-m** [comment]
Try with `-rtsp_transport tcp` for the `source`:
```
{
  "platform": "Camera-ffmpeg",
  "cameras": [
    {
      "name": "Camera Name",
      "videoConfig": {
      	"source": "-rtsp_transport tcp -re -i rtsp://myfancy_rtsp_stream",
        "stillImageSource": "-i http://faster_still_image_grab_url/this_is_optional.jpg",
      	"maxStreams": 2,
      	"maxWidth": 1280,
      	"maxHeight": 720,
      	"maxFPS": 30
      }
    }
  ]
}
```
In the end i'm using https://github.com/legotheboss/homebridge-camera-ffmpeg-omx though, i'm on a Raspberry PI3, so that is the best option for those with that hardware.

**Comment by BigDogJonx** [comment]
@fulvio-m 
Your solution of adding `-rtsp_transport tcp` clearly resolved it.  I also went ahead moved over to the ffmpeg-omx one.  Initially I tried this awhile back and saw no improvement, if anything was actually slightly worse, so came back to this version.  I did however have that ffmpeg installed from that ffmpeg-omx install guide from legotheboss.

Appreciate this bigtime.  Now its fully usable and stream is clean, clear, and realtime.

Dixit
